### ISSUE TITLE
cli.py sync crashes

### ISSUE THREAD

**Issue by IzzySoft** [ya0211/magisk-modules-repo-util#28]
With the new release, running `cli.py sync` crashes when pulling an update:

```
[2023/08/30 00:58:01] Pull DEBUG: _get_changelog_common: [jdtoolbox] -> Add support for using bootable kernel flashers as input to kernel installer
Add kernel+dtb install menu is not in /sanitized/repo/magisk/local
[2023/08/30 00:58:01] Pull DEBUG: from_track: [jefferderp.keyboardremaps] -> type: ONLINE_JSON
Traceback (most recent call last):
  File "util/cli.py", line 26, in <module>
    sys.exit(Main.exec())
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 53, in exec
    code = cls._check_args()
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 73, in _check_args
    return cls.sync()
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 265, in sync
    sync.update(
  File "/sanitized/repo/magisk/util/sync/core/Sync.py", line 141, in update
    online_module = self._update_jsons(track=track, force=force)
  File "/sanitized//repo/magisk/util/sync/core/Sync.py", line 38, in _update_jsons
    online_module, timestamp = self._pull.from_track(track)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 221, in from_track
    return self.from_json(track, local=False)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 167, in from_json
    online_module = self._from_zip_common(track.id, zip_file, changelog, delete_tmp=True)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 124, in _from_zip_common
    changelog_url = self._get_file_url(module_id, target_changelog_file)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 49, in _get_file_url
    if not (file.is_relative_to(module_folder) and file.exists()):
AttributeError: 'PosixPath' object has no attribute 'is_relative_to'
```

This did not happen with the repo-util version from April.

Here's the corresponding `track.json` for this module:

```json
{
  "id": "jefferderp.keyboardremaps",
  "update_to": "https://github.com/Jefferderp/Magisk-KeyboardRemaps/raw/master/update.json",
  "license": "GPL-2.0-only",
  "homepage": "",
  "source": "https://github.com/Jefferderp/Magisk-KeyboardRemaps",
  "support": "https://github.com/Jefferderp/Magisk-KeyboardRemaps/issues",
  "donation": "",
  "changelog": "https://github.com/Jefferderp/Magisk-KeyboardRemaps/raw/master/changelog.md",
  "added": 1679706725.0,
  "last_update": 1647987772.0,
  "versions": 1
}
```

and the corresponding `update.json`:

```json
{
  "id": "jefferderp.keyboardremaps",
  "timestamp": 1647987772.0,
  "versions": [
    {
      "timestamp": 1647987772.0,
      "version": "1",
      "versionCode": 10,
      "zipUrl": "https://apt.izzysoft.de/magisk/modules/jefferderp.keyboardremaps/1_1.zip",
      "changelog": "https://apt.izzysoft.de/magisk/modules/jefferderp.keyboardremaps/1_1.md"
    }
  ]
}
```

Looking into the corresponding module directory: the new ZIP as well as the changelog have been pulled. No other files seem to have been updated (especially neither `update.json` nor `track.json`). There are 2 ZIPs now in the directory. Funnily, both (`1_1.zip` and `v1.0_10.zip`) seem to use `versionCode: 10`. And there was no commit or new release in that repo since April, so repo-util should not even have pulled anything. No idea why the two ZIP files even differ in size (the newer one is smaller).

Removing the original ZIP and MD and running `cli.py sync` again seems to solve it for this module, but then fails for another with a stack trace:

```
[2023/08/30 01:12:45] Pull DEBUG: from_track: [unlimited-hotspot] -> type: LOCAL_ZIP
Traceback (most recent call last):
  File "util/cli.py", line 26, in <module>
    sys.exit(Main.exec())
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 53, in exec
    code = cls._check_args()
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 73, in _check_args
    return cls.sync()
  File "/sanitized/repo/magisk/util/sync/cli/Main.py", line 265, in sync
    sync.update(
  File "/sanitized/repo/magisk/util/sync/core/Sync.py", line 141, in update
    online_module = self._update_jsons(track=track, force=force)
  File "/sanitized/repo/magisk/util/sync/core/Sync.py", line 38, in _update_jsons
    online_module, timestamp = self._pull.from_track(track)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 229, in from_track
    return self.from_zip(track)
  File "/sanitized/repo/magisk/util/sync/core/Pull.py", line 206, in from_zip
    last_modified = zip_file.stat().st_mtime
  File "/usr/lib/python3.8/pathlib.py", line 1198, in stat
    return self._accessor.stat(self)
FileNotFoundError: [Errno 2] No such file or directory: '/sanitized/repo/magisk/local/unlimited_hotspot_v5_5.zip'
```

Again, this is a module available in my repo for quite a while, and `sync` never complained before the update to the new v2.0 release of repo-util. Why does it suddenly look for the ZIP in the `local/` directory? It is properly located in `modules/unlimited_hotspot/v5_5.zip` (a relative path to `local/` would be needed for `LOCAL_JSON` – I see that's now also needed for `LOCAL_ZIP`). With the module's source repo gone, I simply disabled updates for now (`cli.py trrack -i unlimited_hostspot --disable-updates`), which seems to have worked fine. `sync` runs through now – but after `index` I notice one module seems to be gone (count decreased). Digging into it I see this log entry:

```
[2023/08/30 01:26:00] LocalTracks INFO: get_tracks: size = 81
```

But there are only 80 modules in the index. Luckily I kept the previous one: it's our first candidate, `jefferderp.keyboardremaps`, which is missing now; it's `update.json` still had the original file names for `zipUrl` and `changelog` (that wasn't updated when `sync` crashed). Fixing those manually, too, brought the module back.

**TL;DR:** while the second case (`LOCAL_ZIP`) could be declared "works as designed", something is definitely wrong with the first one. Not sure what changed in the handling here between the versions. Further, `sync` should not crash in such a case but rather leave the affected module as.is, just throwing an appropriate warning/error – else automated updates via cron would simply get stuck for the entire repo if a single module fails to sync/update.

I also miss any hints in the log output indicating which modules got updates (there were such entries with the older repo-util).

----

One more strange message:

```
[2023/08/30 01:25:51] Pull DEBUG: _get_changelog_common: [jdtoolbox] -> Add support for using bootable kernel flashers as input to kernel installer
Add kernel+dtb install menu is not in /sanitized/repo/magisk/local
```

Oof. That's an error in the module's repo, see https://raw.githubusercontent.com/JoshuaDoes/jdtoolbox/master/update.json – maybe ignore URLs which do not start with `http` (or any other supported protocol)?

**Comment by IzzySoft** [comment]
Thanks! With this last one being addressed, would you say the current state is stable and I should switch to this (which I assume is the case) – or should I better wait for the next release, @ya0211?
